### PR TITLE
Add support for Goliathus mouse mats

### PR DESCRIPTION
**Pull request by intervigilium**
I suck at scraping the Razer website, so the images are just copies of Firefly right now.

### CODE REVIEW

File: daemon/openrazer_daemon/hardware/mouse_mat.py [hunk 1]
```diff
@@ -24,3 +24,41 @@ class RazerFirefly(__RazerDeviceBrightnessSuspend):
...
+    MATRIX_DIMS = [1, 1]
+    METHODS = ['get_device_type_mousemat', 'set_static_effect', 'set_spectrum_effect',
+               'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect', 'set_breath_dual_effect',
+               'set_custom_effect', 'set_key_row']
+
+    RAZER_URLS = {
+        "top_img": "https://assets.razerzone.com/eeimages/support/products/1394/1394_goliathus_chroma.png",
+        "side_img": "https://assets.razerzone.com/eeimages/products/21936/fireflycloth-gallery-6.png",
+        "perspective_img": "https://assets.razerzone.com/eeimages/products/21936/fireflycloth-gallery-3.png"
+    }
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse_mat.py]: Remove those images and replace them with the `DEVICE_IMAGE` as added on master a few days ago (see @lah7 s suggestion)

File: daemon/openrazer_daemon/hardware/mouse_mat.py [hunk 2]
```diff
@@ -24,3 +24,41 @@ class RazerFirefly(__RazerDeviceBrightnessSuspend):
...
+    MATRIX_DIMS = [1, 1]
+    METHODS = ['get_device_type_mousemat', 'set_static_effect', 'set_spectrum_effect',
+               'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect', 'set_breath_dual_effect',
+               'set_custom_effect', 'set_key_row']
+
+    RAZER_URLS = {
+        "top_img": "https://assets.razerzone.com/eeimages/support/products/1395/1395_goliathusextended.png",
+        "side_img": "https://assets.razerzone.com/eeimages/products/21936/fireflycloth-gallery-6.png",
+        "perspective_img": "https://assets.razerzone.com/eeimages/products/21936/fireflycloth-gallery-3.png"
+    }
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse_mat.py]: same here

File: daemon/openrazer_daemon/hardware/mouse_mat.py [hunk 3]
```diff
@@ -24,3 +24,41 @@ class RazerFirefly(__RazerDeviceBrightnessSuspend):
...
+
+
+class RazerGoliathus(__RazerDeviceBrightnessSuspend):
+    """
+    Class for the Razer Goliathus (2018)
+    """
+    USB_VID = 0x1532
+    USB_PID = 0x0C01
+    HAS_MATRIX = True
+    MATRIX_DIMS = [1, 1]
```
Comment (z3ntu) [inline daemon/openrazer_daemon/hardware/mouse_mat.py:36]: `[1, 1]`? Sounds a bit wrong

File: driver/razercommon.h [hunk 4]
```diff
@@ -40,6 +40,7 @@ do { \
 #define VARSTORE         0x01
 
 // LED definitions
+#define MOUSEPAD_LED      0x00
```
Comment (z3ntu) [inline driver/razercommon.h]: Maybe call it `ZERO_LED` or something else but this LED is also used for other devices so let's not make it "mousepad"-specific

File: driver/razermousemat_driver.c [hunk 5]
```diff
@@ -40,12 +40,14 @@ MODULE_DESCRIPTION(DRIVER_DESC);
...
 MODULE_LICENSE(DRIVER_LICENSE);
 
+static unsigned char saved_brightness = 0;
 
 /**
  * Send report to the mousemat
  */
 int razer_get_report(struct usb_device *usb_dev, struct razer_report *request_report, struct razer_report *response_report)
 {
+    // TODO: Rename WAIT_MIN_US from FIREFLY
```
Comment (z3ntu) [inline driver/razermousemat_driver.c]: solve that todo please, should be a simple find&replace

File: driver/razermousemat_driver.c [hunk 6]
```diff
@@ -93,11 +95,23 @@ static ssize_t razer_attr_read_set_brightness(struct device *dev, struct device_
     struct usb_interface *intf = to_usb_interface(dev->parent);
     struct usb_device *usb_dev = interface_to_usbdev(intf);
     struct razer_report response;
-    struct razer_report report = razer_chroma_standard_get_led_brightness(VARSTORE, BACKLIGHT_LED);;
+    struct razer_report report = {0};
```
Comment (z3ntu) [inline driver/razermousemat_driver.c:97]: Why not just `struct razer_report report;`? I see that the report doesn't get used in the new codepath. If you think this is better like this, please also update the rest of the occurences of `struct razer_report report;` to `= {0};` in a separate commit

**Review by z3ntu — COMMENTED** [review]
Pretty good, see comments please. 
Also are you sure the device doesn't have a "get brightness" method we can use? Often the Razer firmwares support some features even though Synapse never uses them so we might have luck here

File: driver/razermousemat_driver.c (hunk 6 above)
Comment (z3ntu) [reply driver/razermousemat_driver.c:97]: With the command `git grep "struct razer_report" | grep -v '=' | grep ';'` all instances should be found I think

File: daemon/openrazer_daemon/hardware/mouse_mat.py [hunk 7]
```diff
@@ -24,3 +24,41 @@ class RazerFirefly(__RazerDeviceBrightnessSuspend):
...
+    METHODS = ['get_device_type_mousemat', 'set_static_effect', 'set_spectrum_effect',
+               'set_none_effect', 'set_breath_random_effect', 'set_breath_single_effect', 'set_breath_dual_effect',
+               'set_custom_effect', 'set_key_row']
+
+    RAZER_URLS = {
```
Comment (lah7) [inline daemon/openrazer_daemon/hardware/mouse_mat.py]: Can be amended in line with aeabb468fe1ec700ac4cc1647cd09b93a26daeed like so:

```suggestion
    DEVICE_IMAGE = "https://assets.razerzone.com/eeimages/support/products/1394/1394_goliathus_chroma.png"

    # Deprecated - RAZER_URLS be removed in future.    
    RAZER_URLS = {
```

**Comment by intervigilium** [comment]
I couldn't find a working get brightness function. The extended_matrix variant that would normally be used is not valid.

File: driver/razermousemat_driver.c (hunk 6 above)
Comment (intervigilium) [reply driver/razermousemat_driver.c:97]: Using = {0}; is a zero initialization of the struct. Some compilers warn if variables are left uninitialized. Should be easy enough to update separately.

File: daemon/openrazer_daemon/hardware/mouse_mat.py (hunk 3 above)
Comment (intervigilium) [reply daemon/openrazer_daemon/hardware/mouse_mat.py:36]: My understanding is that this is for custom matrix commands. Goliathus allows such commands, but you only get 1 addressable LED...

File: driver/razermousemat_driver.c (hunk 5 above)
Comment (intervigilium) [reply driver/razermousemat_driver.c]: Done

File: driver/razercommon.h (hunk 4 above)
Comment (intervigilium) [reply driver/razercommon.h]: Renamed to ZERO_LED.

File: daemon/openrazer_daemon/hardware/mouse_mat.py (hunk 1 above)
Comment (intervigilium) [reply daemon/openrazer_daemon/hardware/mouse_mat.py]: Removed and replaced with DEVICE_IMAGE

File: daemon/openrazer_daemon/hardware/mouse_mat.py (hunk 2 above)
Comment (intervigilium) [reply daemon/openrazer_daemon/hardware/mouse_mat.py]: Removed and replaced with DEVICE_IMAGE

File: driver/razermousemat_driver.c (hunk 6 above)
Comment (intervigilium) [reply driver/razermousemat_driver.c:97]: Added a follow-up patch to find and replace.

File: daemon/openrazer_daemon/hardware/mouse_mat.py (hunk 7 above)
Comment (intervigilium) [reply daemon/openrazer_daemon/hardware/mouse_mat.py]: I just replaced these with DEVICE_IMAGE.

**Comment by intervigilium** [comment]
Anything left here?

**Comment by z3ntu** [comment]
I've taken the `driver: Zero-initialize razer_report struct` commit, expanded it with more occurrences and pushed it to master now.

I think after rebasing on master it should be good to merge

**Comment by intervigilium** [comment]
Rebased it.

**Review by terricain — APPROVED** [review]

**Review by z3ntu — APPROVED** [review]
LGTM, thanks!

**Pull request merged**